src/java/org/apache/cassandra/cql3/statements/ModificationStatement.java

```diff
@@ -816,9 +818,25 @@ final void addUpdates(UpdatesCollector collector,
                 }
                 else
                 {
+                    // Clustering keys need to be checked on their own
                     for (Clustering<?> clustering : clusterings)
                     {
                         clustering.validate();
+                        for (int i = 0; i < clustering.size(); i++)
+                        {
+                            ColumnMetadata column = metadata.clusteringColumns().get(i);
+                            if (column.hasConstraint())
+                            {
+                                try
+                                {
+                                    clustering.checkConstraints(i, metadata.comparator, column.getColumnConstraints());
+                                }
+                                catch (ConstraintViolationException e)
+                                {
+                                    throw new InvalidRequestException(e.getMessage(), e);
+                                }
+                            }
+                        }
```

Comment:
nit: extract out to a dedicated method, since the method `addUpdates` is quite large already.